core/arch/arm/plat-stm32mp1/drivers/stm32mp1_rcc.c

```diff
@@ -45,28 +33,46 @@ static uint8_t reset_id2reg_bit_pos(unsigned int reset_id)
 	return reset_id & RESET_BIT_POS_MASK;
 }
 
-void stm32_reset_assert(unsigned int id)
+TEE_Result stm32_reset_assert(unsigned int id, unsigned int to_us)
 {
 	size_t offset = reset_id2reg_offset(id);
 	uint32_t bitmsk = BIT(reset_id2reg_bit_pos(id));
-	uint64_t timeout_ref = timeout_init_us(RESET_TIMEOUT_US);
 	vaddr_t rcc_base = stm32_rcc_base();
 
 	io_write32(rcc_base + offset, bitmsk);
 
-	WAIT_COND_OR_PANIC(io_read32(rcc_base + offset) & bitmsk,
-			   timeout_ref);
+	if (to_us) {
+		uint64_t timeout_ref = timeout_init_us(to_us);
+
+		while (!(io_read32(rcc_base + offset) & bitmsk))
+			if (timeout_elapsed(timeout_ref))
+				break;
+
+		if (!(io_read32(rcc_base + offset) & bitmsk))
```

Comment:
This seems redundant, we've just done it above in the while loop.
Same pattern below in `stm32_reset_deassert()`.

Reply:
Consider `stm32_reset_assert()` (and `stm32_reset_deassert()`) being called from a threaded context. While looping at lines 47/48, TEE thread maybe interrupted and rescheduled after timeout occurs. In such case, we don't consider the timeout as far as SoC register shows the target reset level is reached.

Reply:
With that logic you could argue that we should test it three times just in case.

Reply:
Sorry I don't understand your last comment.

Macro `WAIT_COND_OR_PANIC()` (before this change) made such a double test: timeout was considered only if reset signal did not reach target state.

This change removes the macro as we expect to return an error instead of panicking and I think using a `return TEE_ERROR_xxx;` from a macro is not really clean. Therefore I preferred to use 2 explicit sequences rather than changing `WAIT_COND_OR_PANIC()` into something like:
```c
#define WAIT_COND_OR_ERROR(_condition, _timeout_ref) \	
	do { \	
		if (timeout_elapsed(_timeout_ref) && !(_condition)) \	
			return TEE_ERROR_SECURITY; \	
	} while (!(_condition))
```

Reply:
For the sake of the argument assume that the timeout corresponds to 10000 rounds in the loop. That means that we need to check the device 10000 times until we have reached the timeout. Once we've reached the timeout we check the device for the 10001th time in case the state should have changed even though we're already on overtime. Does that make sense?

Reply:
Know I get it, sorry. This OK.

Reply:
ok, thanks for the time spent checking whether this makes sense.